### PR TITLE
Add heartbeat-based dynamic profiling system to intel repo

### PR DESCRIPTION
**Pull request by ashokbytebytego**
# Add Heartbeat-Based Dynamic Profiling System

## 🎯 Overview

This PR introduces a heartbeat protocol that enables dynamic profiling control. Agents periodically send heartbeats to a Performance Studio backend and receive start/stop profiling commands, allowing on-demand profiling without agent restarts.

## 💡 Motivation

Current gProfiler deployment requires:
- Profilers to run continuously, consuming resources even when profiling isn't needed
- Agent restarts to change profiling configuration
- Manual coordination for targeted profiling campaigns

This heartbeat system solves these issues by enabling:
- **Centralized control**: Manage all agents from a single Performance Studio backend
- **On-demand profiling**: Profile only when needed, reducing resource usage
- **Dynamic configuration**: Change profiler settings without agent restarts
- **Idempotent execution**: Prevent duplicate work through UUID-based command tracking

## 📦 Changes Summary

**5 files changed**: 4 new files, 1 modified  
**Total changes**: 1,838 insertions, 30 deletions

#### 1. `gprofiler/heartbeat.py` (627 lines)
**Core heartbeat system implementation**

- **`HeartbeatClient`** class:
  - Sends periodic heartbeats to Performance Studio backend
  - Receives start/stop profiling commands from server
  - Reports command completion status
  - Tracks executed command IDs for idempotency (in-memory with configurable limit)
  - Handles authentication via Bearer token
  - SLI metrics integration for success/failure tracking

- **`DynamicGProfilerManager`** class:
  - Manages dynamic profiler lifecycle (start/stop based on server commands)
  - Creates and configures GProfiler instances dynamically
  - Handles command types: `start` and `stop`
  - Supports dynamic profiler configuration:
    - Duration, frequency, profiling mode
    - Target hostnames and PIDs
    - Per-profiler settings (Perf, PyPerf, PySppy, Java, PHP, Ruby, .NET, NodeJS)
    - PerfSpect hardware metrics integration
    - Max processes per profiler
  - Ensures proper cleanup of profiler resources
  - Thread-based profiler execution with graceful shutdown

**Key Features**:
- Command idempotency prevents duplicate execution
- Comprehensive error handling and logging
- Memory-efficient command history tracking (max 1000 commands)
- Automatic subprocess cleanup after profiling

#### 2. `docs/HEARTBEAT_SYSTEM_README.md` (634 lines)
**Complete system documentation**

Contains:
- System architecture overview with diagrams
- API endpoint specifications (3 endpoints):
  - `POST /api/metrics/profile_request` - Submit profiling requests
  - `POST /api/metrics/heartbeat` - Agent heartbeat
  - `POST /api/metrics/command_completion` - Report command completion
- Backend and agent features documentation
- PerfSpect hardware metrics integration guide
- Usage examples (command-line, curl examples)
- Configuration options
- Testing instructions (mock mode and live mode)
- Troubleshooting guide
- Building and running locally
- Security considerations

#### 3. `tests/test_heartbeat_system.py` (358 lines)
**Comprehensive test suite**

- **Mock mode** (default): No backend required
  - Uses `unittest.mock` to simulate backend responses
  - Perfect for CI/CD and quick testing
  - Tests all heartbeat flows without external dependencies

- **Live mode** (`--live` flag): Tests with real backend
  - Verifies integration with actual Performance Studio backend
  - Useful for end-to-end testing

**Test Coverage**:
- ✅ Initial heartbeat (no commands)
- ✅ CREATE and EXECUTE start command
- ✅ Command idempotency (duplicate command handling)
- ✅ CREATE and EXECUTE stop command  
- ✅ Multiple heartbeats with no pending commands
- ✅ Command completion acknowledgments
- ✅ Error handling and edge cases

**Classes & Functions**:
- `HeartbeatClient`: Simulates agent behavior
- `create_test_profiling_request()`: Submit test requests
- `create_mock_responses()`: Mock backend for testing
- `run_tests()`: Execute full test suite
- `main()`: Entry point with mode selection

#### 4. `tests/run_heartbeat_agent.py` (136 lines)
**Agent test runner**

- Demonstrates how to run gProfiler in heartbeat mode
- Configurable test environment
- Helpful for local development and testing
- Shows proper command-line argument usage
- Includes instructions and usage guide

#### 1. `gprofiler/main.py` (+83 lines, -30 lines)
**Heartbeat mode integration**

**Changes**:

1. **Import** (line 51):
   ```python
   from gprofiler.heartbeat import DynamicGProfilerManager, HeartbeatClient
   ```

2. **CLI Arguments** (lines 865-881):
   ```python
   --enable-heartbeat-server      # Enable heartbeat mode
   --heartbeat-interval <seconds> # Heartbeat frequency (default: 30s)
   ```

3. **Argument Validation** (lines 956-964):
   - Validates `--enable-heartbeat-server` requires `--upload-results`
   - Validates authentication token is provided
   - Validates service name is provided

4. **Heartbeat Mode Execution** (lines 1244-1274):
   - Checks if heartbeat mode is enabled
   - Creates `HeartbeatClient` with server connection details
   - Creates `DynamicGProfilerManager` to manage profiler lifecycle
   - Starts heartbeat loop (waits for server commands)
   - Falls back to normal profiling mode if heartbeat is disabled
   - Proper error handling and graceful shutdown

### Agent Features
✅ Heartbeat communication with configurable intervals (default: 30s)  
✅ Dynamic profiling based on server commands (start/stop)  
✅ Command-driven execution with full configuration control  
✅ Idempotency to prevent duplicate command execution  
✅ In-memory command tracking (max 1000 commands)  
✅ Graceful error handling and retry logic  
✅ PerfSpect hardware metrics support (auto-installation in dynamic mode)  
✅ SLI metrics integration for monitoring  
✅ Comprehensive subprocess cleanup  

### Backend Features (Expected from Performance Studio)
✅ REST API for submitting profiling requests  
✅ Heartbeat endpoint for agent communication  
✅ Command merging for multiple requests targeting same host  
✅ Process-level and host-level stop commands  
✅ Idempotent command execution using unique command IDs  
✅ Command completion tracking  
✅ PerfSpect integration for hardware metrics  

### Configuration Options
The heartbeat system supports dynamic configuration of:
- Profiling duration and frequency
- Profiling mode (CPU, allocation, none)
- Target hostnames and PIDs
- Per-profiler configuration:
  - **Perf**: enabled_restricted / enabled_aggressive / disabled
  - **PyPerf**: enabled / disabled
  - **PySppy**: enabled_fallback / enabled / disabled
  - **Java async-profiler**: enabled / disabled
  - **PHP**: enabled / disabled
  - **Ruby**: enabled / disabled
  - **.NET**: enabled / disabled
  - **NodeJS**: enabled / disabled
- PerfSpect hardware metrics: enabled / disabled
- Max processes per profiler (default: 10)
- Continuous vs single-shot profiling

### Starting Agent in Heartbeat Mode

```bash
python gprofiler/main.py \
  --enable-heartbeat-server \
  --upload-results \
  --token "your-token" \
  --service-name "web-service" \
  --api-server "http://performance-studio:8000" \
  --heartbeat-interval 30 \
  --output-dir /tmp/profiles \
  --verbose
```

### Backend: Submit Start Command

```bash
curl -X POST http://backend:8000/api/metrics/profile_request \
  -H "Content-Type: application/json" \
  -d '{
    "service_name": "web-service",
    "command_type": "start",
    "duration": 60,
    "frequency": 11,
    "profiling_mode": "cpu",
    "target_hostnames": ["host1", "host2"],
    "additional_args": {
      "enable_perfspect": true,
      "max_processes": 10,
      "profiler_configs": {
        "perf": "enabled_restricted",
        "pyperf": "enabled",
        "pyspy": "enabled_fallback"
      }
    }
  }'
```

### Backend: Submit Stop Command

```bash
curl -X POST http://backend:8000/api/metrics/profile_request \
  -H "Content-Type: application/json" \
  -d '{
    "service_name": "web-service",
    "command_type": "stop",
    "stop_level": "host",
    "target_hostnames": ["host1"]
  }'
```

### Run Mock Tests (No Backend Required)
```bash
python tests/test_heartbeat_system.py
```

Expected output:
```
🧪 Testing Heartbeat-Based Profiling Control System
🎭 Running in MOCK MODE (no real backend required)
============================================================
✓ All tests passed!

Test Summary:
   - Executed commands: 2
   - Commands executed: ['cmd_1', 'cmd_2']

📊 Mock Backend State:
   - Total heartbeats: 6
   - Completed commands: 2
```

### Run Live Tests (Requires Backend)
```bash
python tests/test_heartbeat_system.py --live
```

### Run Agent Test Runner
```bash
python tests/run_heartbeat_agent.py
```

### System Flow

```
┌─────────────────┐                    ┌─────────────────┐
│  Performance    │                    │   gProfiler     │
│  Studio Backend │                    │     Agent       │
└────────┬────────┘                    └────────┬────────┘
         │                                      │
         │◄──── Heartbeat (every 30s) ─────────┤
         │      {hostname, service, status}     │
         │                                      │
         ├────── Start Command ────────────────►│
         │      {duration: 60, frequency: 11}   │
         │                                      │
         │                               [Profiling...]
         │                                      │
         │◄──── Command Completion ─────────────┤
         │      {status: completed, time: 65s}  │
         │                                      │
         │◄──── Heartbeat ──────────────────────┤
         │                                      │
         ├────── Stop Command ─────────────────►│
         │                                      │
         │◄──── Command Completion ─────────────┤
         │      {status: completed}             │
```

### Command Lifecycle

1. **User** submits profiling request to Performance Studio
2. **Backend** creates command with unique UUID
3. **Agent** sends heartbeat to backend
4. **Backend** responds with pending command
5. **Agent** checks command hasn't been executed (idempotency)
6. **Agent** executes command (start profiler with config)
7. **Agent** reports completion to backend
8. **Backend** updates command status

## 🛡️ Dependencies

**No new dependencies required!** ✅

All dependencies used by the heartbeat system already exist in the project:
- `requests` - Already in requirements.txt
- `configargparse` - Already used in main.py
- `psutil` - Already in requirements.txt
- `threading` - Python standard library
- `datetime` - Python standard library
- `socket` - Python standard library

## ✅ Compatibility

- ✅ Works with all existing gProfiler profilers (Perf, PyPerf, Java, PHP, Ruby, .NET, NodeJS)
- ✅ Compatible with existing output mechanisms (local files, server upload)
- ✅ Preserves all existing gProfiler features when not in heartbeat mode
- ✅ Based on Intel's latest `dynamic_profiling` branch (commit 37e6acc)
- ✅ No breaking changes to existing functionality
- ✅ Backward compatible with existing deployments

## 📚 Documentation

Complete documentation available in:
- `docs/HEARTBEAT_SYSTEM_README.md` - Full system documentation
- Code comments in `gprofiler/heartbeat.py` - Implementation details
- Test files - Usage examples and test scenarios

### CODE REVIEW

File: gprofiler/heartbeat.py [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+if TYPE_CHECKING:
+    from gprofiler.main import GProfiler
+
+from gprofiler.client import ProfilerAPIClient
+from gprofiler.containers_client import ContainerNamesClient
+from gprofiler.metadata.application_identifiers import ApplicationIdentifiers
+from gprofiler.metadata.enrichment import EnrichmentOptions
+from gprofiler.metadata.metadata_collector import get_static_metadata
+from gprofiler.metadata.system_metadata import get_hostname
+from gprofiler.metrics_publisher import (
```
Comment (mlim19) [inline gprofiler/heartbeat.py]: metrics_publisher module seems missing in this PR. please check.

File: gprofiler/heartbeat.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+                )
+                
+                if result.get("success") and result.get("profiling_command"):
+                    logger.info(f"Received profiling command from server: {result.get('command_id')}")
+                    return result
+                else:
+                    logger.debug("Heartbeat successful, no pending commands")
+                    return None
+            else:
+                logger.warning(f"Heartbeat failed with status {response.status_code}: {response.text}")
```
Comment (mlim19) [inline gprofiler/heartbeat.py]: This seems to be treated as error as well, isn't it?

File: gprofiler/heartbeat.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+                # Emit success metric (SLI tracking) using singleton
+                MetricsPublisher.get_instance().send_sli_metric(
+                    response_type=RESPONSE_TYPE_SUCCESS,
+                    method_name='send_heartbeat'
+                )
+                
+                if result.get("success") and result.get("profiling_command"):
+                    logger.info(f"Received profiling command from server: {result.get('command_id')}")
+                    return result
+                else:
```
Comment (mlim19) [inline gprofiler/heartbeat.py]: This else may handle 2 cases: case1 (not success) and case2 (success but no profiling_command). The first case shouldn't be handled as heartbeat successful. Please handle them separately

File: gprofiler/heartbeat.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+            try:
+                # Send heartbeat and check for commands
+                command_response = self.heartbeat_client.send_heartbeat()
+                
+                if command_response and command_response.get("profiling_command"):
+                    profiling_command = command_response["profiling_command"]
+                    command_id = command_response["command_id"]
+                    command_type = profiling_command.get("command_type", "start")
+                    
+                    logger.info(f"Received profiling command: {profiling_command}")
```
Comment (mlim19) [inline gprofiler/heartbeat.py]: This is redundant with the line 119. It would be better to turn one of them into debug

File: gprofiler/heartbeat.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+
+                        # Wait for next heartbeat
+                        self.stop_event.wait(self.heartbeat_interval)
+
+                        continue
+                    
+                    logger.info(f"Received {command_type} command: {command_id}")
+                    
+                    # Mark command as executed for idempotency
+                    self.heartbeat_client.mark_command_executed(command_id)
```
Comment (mlim19) [inline gprofiler/heartbeat.py]: I think these status update should happen only for valid command type.

File: gprofiler/heartbeat.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+                            status="completed",
+                            execution_time=0,
+                            error_message=None,
+                            results_path=None
+                        )
+                    elif command_type == "start":
+                        # Stop current profiler if running, then start new one
+                        logger.info("Starting new profiler due to start command")
+                        # TODO: important comment to make sure profiler has stopped successful to avoid leak 
+                        self._stop_current_profiler()
```
Comment (mlim19) [inline gprofiler/heartbeat.py]: Please add the error handling for the cases stop and start are not processed properly.

File: gprofiler/heartbeat.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+                daemon=True
+            )
+            self.current_thread.start()
+            
+            logger.info(f"Started new gProfiler instance with command ID: {command_id}")
+            
+        except Exception as e:
+            logger.error(f"Failed to start new profiler: {e}", exc_info=True)
+            # Report failure to the server
+            self.heartbeat_client.send_command_completion(
```
Comment (mlim19) [inline gprofiler/heartbeat.py]: This code might send the duplicated command completion message when exception is raised. Please remove it if it's right.

**Comment by mlim19** [comment]
Please fix build and linter issues.

File: gprofiler/main.py [hunk 8]
```diff
@@ -1215,37 +1240,65 @@ def main() -> None:
...
-            perfspect_duration=getattr(args, "tool_perfspect_duration", 60),
-            verbose=args.verbose,
-        )
-        logger.info("gProfiler initialized and ready to start profiling")
-        if args.continuous:
-            gprofiler.run_continuous()
+        
+        # Check if heartbeat server mode is enabled FIRST
+        if args.enable_heartbeat_server:
+            if not args.upload_results:
```
[the remark this answers is not in the document]
Comment (ashokbytebytego) [reply gprofiler/main.py]: Removed the duplicate check

File: gprofiler/heartbeat.py (hunk 3 above)
Comment (ashokbytebytego) [reply gprofiler/heartbeat.py]: Separated logic

**Comment by ashokbytebytego** [comment]
## Testing 

Tested by building gProfiler agent locally and verified:
- Agent connects to Performance Studio backend successfully
- Heartbeat communication working as expected
- Start/stop commands received and executed properly
- Profiles generated and uploaded to backend
- Command completion status reported correctly
- Error handling and logging functioning as designed

**Comment by ashokbytebytego** [comment]
> > @ashokbytebytego and @prashantbytesyntax, I just tested this PR with the gprofiler-performance-studio change. The typical gprofiler collection (without heartbeat mode) shows the following error:
> > ```
> > **[2026-01-12 18:30:37,792] CRITICAL: gprofiler.profilers.perf: Failed to determine perf event to use
> > [2026-01-12 18:30:37,793] CRITICAL: gprofiler.profilers.factory: Couldn't create the Perf profiler, not continuing. Run with --no-perf to disable this profiler
> > Traceback (most recent call last):
> >   File "gprofiler/profilers/factory.py", line 54, in get_profilers
> >   File "gprofiler/profilers/perf.py", line 203, in __init__
> >   File "gprofiler/utils/perf.py", line 118, in discover_appropriate_perf_event
> > gprofiler.exceptions.PerfNoSupportedEvent**
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It seems the function "discover_appropriate_perf_event" in perf.py is not working as expected. Can you please check?
> 
> Hi @mlim19 , Please use the -enable-heartbeat-server option when starting the agent to test dynamic profiling. I can then investigate why it's breaking without dynamic profiling enabled. Thank you cc: @prashantbytesyntax


Hi @mlim19 , I did tried with out enabling --enable-heartbeat-server and used below command to run gprofiler. I was not able to reproduce issue.  this might be perf compatibility issue on your machine. 

sudo TMPDIR=/proc/self/cwd sh -c "setsid ./gprofiler -cu --token='ENfzQ_4YjXXxA_7C6GRv6ACcpT2wAou3SfmuYoGnYvA' --service-name='test_dynamic_profiling' --server-host "https://localhost:4433/" --glogger-server "https://localhost:4433/" --no-verify"

**Review by artursarlo — APPROVED** [review]
Tested locally with compatible backend implementation. Working as expected.
